File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left=-1,right=-1,sum=0;
        for(int w:weights){
            sum+=w;
            left=Math.max(left,w);
        }
        right=left*(weights.length/days+Integer.signum(weights.length%days));
        left=Math.max(left,sum/days);
        while (left<right){
            int mid=left+(right-left)/2;
            if(good(mid,weights,days)) right=mid;
            else left=mid+1;
        }
        return left;
    }
    public boolean good(int res, int[] weights, int days){
        int sum=0;
        for (int w:weights){
            if(sum+w>res){
                sum=0;
                days--;
            }
            sum+=w;
        }
        return days>0;
    }
}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left,right = max(weights),sum(weights)
        while left < right:
            mid = left + (right - left)//2
            daysNeed, currWeight = 1,0
            for weight in weights:
                if currWeight + weight > mid:
                    daysNeed += 1
                    currWeight = 0
                currWeight += weight
            if daysNeed > days:
                left = mid + 1
            else: 
                right = mid
        return left